space_packet_parser/xtce/definitions.py

```diff
@@ -107,207 +288,85 @@ def _populate_parameter_type_cache(self):
-            custom_algorithm_element = restriction_criteria_element.find('xtce:CustomAlgorithm', self.ns)
-            if custom_algorithm_element is not None:
-                raise NotImplementedError("Detected a CustomAlgorithm in a RestrictionCriteria element. "
-                                          "This is not implemented.")
-
-            if comparison_list_element is not None:
-                comparison_items = comparison_list_element.findall('xtce:Comparison', self.ns)
-                restrictions = [
-                    comparisons.Comparison.from_match_criteria_xml_element(comp, self.ns) for comp in comparison_items]
-            elif single_comparison_element is not None:
-                restrictions = [
-                    comparisons.Comparison.from_match_criteria_xml_element(single_comparison_element, self.ns)]
-            elif boolean_expression_element is not None:
-                restrictions = [
-                    comparisons.BooleanExpression.from_match_criteria_xml_element(boolean_expression_element, self.ns)]
-            else:
-                raise ValueError("Detected a RestrictionCriteria element containing no "
-                                 "Comparison, ComparisonList, BooleanExpression or CustomAlgorithm.")
-            # TODO: Implement NextContainer support inside RestrictionCriteria. This might make the parser much
-            #    more complicated.
+        parameter_dict = {}
+        if ns:
+            xtce_label, xtce_uri = next(iter(ns.items()))
         else:
-            restrictions = []
-        return self._find_container(base_container_element.attrib['containerRef']), restrictions
+            xtce_uri = ""
+        xtce = f"{{{xtce_uri}}}"
+        for parameter_element in tree.getroot().iterfind(
+                f'{xtce}TelemetryMetaData/{xtce}ParameterSet/{xtce}Parameter', ns):
+            parameter_object = parameters.Parameter.from_xml(parameter_element, ns=ns,
+                                                             parameter_type_lookup=parameter_type_lookup)
+            if parameter_object.name in parameter_dict:
+                raise ValueError(f"Found duplicate parameter name {parameter_object.name}. "
+                                 f"Parameters are expected to be unique")
+            parameter_dict[parameter_object.name] = parameter_object  # Add to cache
+
+        return parameter_dict
+
+    def get_containers(self, *names: str) -> Union[
+        containers.SequenceContainer,
+        Iterable[containers.SequenceContainer]
+    ]:
+        """Get one or more items from the dict cache of SequenceContainer objects"""
+        if len(names) == 1:
+            return self._sequence_container_cache[names[0]]
+        if len(names) == 0:
+            return self._sequence_container_cache
+        return (self._sequence_container_cache[name] for name in names)
```

Comment:
From an API perspective I think I would personally lean towards exposing the underlying dictionaries for the user to call with the direct names: `definition.containers`, `definition.parameters`, `definition.parameter_types`.

There is a lot going on in these right now when I don't think I saw any code asking for multiple names or 0 names leading to the returned cache itself. All of this in one method is confusing IMO. If you prefer methods I think a singular version with only one item returned would be the easiest to understand `get_container()`. But that would also return a `KeyError` for a missing item, so the dictionary access suggestion above might make that more obvious...